### ISSUE TITLE
ecmascript@0.6.2 breaks the compiler

### ISSUE THREAD

**Issue by zvictor** [meteor/meteor#8432]
**Edited**: Reproducible added. It can be seen at [zvictor/meteor-bug-8432](http://github.com/zvictor/meteor-bug-8432).

----

While trying to update Meteor from the version 1.4.2.5 to newer ones, I started seeing `Cannot read property 'name' of undefined` errors.

After many trial and error attempts I could confirm that the problem was caused by the update of the `ecmascript` package. Updating it from previous versions to `ecmascript@0.6.2` raises the error, even in Meteor versions that used to work fine, like the 1.4.2.5.

I started suspecting that it could be related to the issue #8382, as my code makes use of the `__esModule` property. However, removing it from the code didn't make a difference.

I will be working on a reproducible of this error, but I don't have one yet.

The message is as follow:

```haskell
=> Errors prevented startup:

   While processing files with ecmascript (for target web.browser):

   /Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/reify/lib/compiler.js:538:24:
   Cannot read property 'name' of undefined
   at
   /Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/reify/lib/compiler.js:538:24
   at Array.forEach (native)
   at computeSpecifierMap
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/reify/lib/compiler.js:537:14)
   at Context.types.PathVisitor.fromMethodsObject.visitExportNamedDeclaration
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/reify/lib/compiler.js:451:26)
   at Context.invokeVisitorMethod
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:344:49)
   at Visitor.PVp.visitWithoutReset
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:196:32)
   at NodePath.each
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path.js:101:26)
   at visitChildren
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:219:18)
   at Visitor.PVp.visitWithoutReset
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:204:20)
   at visitChildren
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:246:25)
   at Visitor.PVp.visitWithoutReset
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:204:20)
   at visitChildren
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:246:25)
   at Visitor.PVp.visitWithoutReset
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:204:20)
   at Visitor.PVp.visit
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/ast-types/lib/path-visitor.js:133:29)
   at Object.compile
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/reify/lib/compiler.js:29:23)
   at compile
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:52:35)
   at Cache.Cp.get
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/cache.js:94:19)
   at Object.compile
   (/Users/zvictor/.meteor/packages/ecmascript/.0.6.2.1jxhwfk++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:48:23)
   at Object.Babel.compile (packages/babel-compiler.js:48:24)
   at packages/babel-compiler.js:162:22
   at Function.time (/tools/tool-env/profile.js:305:10)
   at profile (packages/babel-compiler.js:192:20)
   at BabelCompiler.BCp.processOneFileForTarget (packages/babel-compiler.js:161:20)
   at BabelCompiler.<anonymous> (packages/babel-compiler.js:93:26)
   at Array.forEach (native)
   at BabelCompiler.BCp.processFilesForTarget (packages/babel-compiler.js:92:14)
```

**Comment by abernix** [comment]
> While trying to update Meteor from the version 1.4.2.5 to newer ones,

_Which_ version of Meteor did you try updating to?  It's important to note specifically which versions you're having issues with when opening an issue.  

As to the problem, if you're using Meteor 1.4.2.x it will _not_ work with `ecmascript@0.6.2` (or higher) due to a breaking API change.  You should either:

* Try upgrading to Meteor 1.4.3.1 and `ecmascript@0.6.3`
* Use Meteor 1.4.2.7 and specific `ecmascript@=0.6.1` in your `.meteor/packages` file to pin to 0.6.1.

I would probably avoid using Meteor 1.4.2.5 at all.  See the `History.md` for [more details](https://github.com/meteor/meteor/blob/devel/History.md#v1427-2017-02-13).

**Comment by zvictor** [comment]
> Which version of Meteor did you try updating to?

I had tried `1.4.2.6`, `1.4.2.7`, `1.4.3` and `1.4.3.1`. I didn't mention the versions before because in not all of them I had managed to force the version of the `ecmascript` package (I was not doing it correctly). All of them were throwing the error mentioned in the first post.


> As to the problem, if you're using Meteor 1.4.2.x it will not work with ecmascript@0.6.2 (or higher). 

How come? Without version pin Meteor 1.4.2.7 installs `ecmascript@0.6.3` by default in my project. 
I am confused.
> Use Meteor 1.4.2.7 and specific ecmascript@=0.6.1 in your .meteor/packages file to pin to 0.6.1.

It works with `ecmascript@=0.6.1`. It fails with `ecmascript@=0.6.2` throwing `Cannot read property 'name' of undefined`.

> Try upgrading to Meteor 1.4.3.1 and ecmascript@0.6.3

It fails throwing `Cannot read property 'name' of undefined`, as in the first post.  And downgrading ecmascript was not allowed by this Meteor version.

**Comment by abernix** [comment]
> How come? Without version pin Meteor 1.4.2.7 installs ecmascript@0.6.3 by default in my project.
> I am confused.

`ecmascript@0.6.3` is just the latest version of `ecmascript` and Meteor 1.4.2.x will automatically install the highest major version of a package that is available due to version-unpinning which happened earlier in Meteor 1.4.  This was intended to make it easier to rapidly release updates to core packages, however it was a bit too free.  As a middle-ground, the flexibility of constraints has been tightened up in Meteor 1.4.3 so it will only take versions from the same minor version (i.e. patch compatible).  See [`History.md`](https://github.com/meteor/meteor/blob/devel/History.md#v143-2017-02-13) for the record of this.

There was a change to the `ecmascript` package which was not compatible with Meteor 1.4.2.x and due to the aforementioned version unpinning in Meteor 1.4.2.x, `ecmascript@0.6.3` is being installed on new projects which are created with Meteor 1.4.2.7 (`meteor create --release 1.4.2.7`), thus the current recommendation to pin to `ecmascript@=0.6.1`.  We'd considered releasing a Meteor 1.4.2.8 which rolls back the changes to `ecmascript` as `0.6.4` (which would be the same as `0.6.2`) and releasing a Meteor 1.4.3.2 with `ecmascript@0.7.0`.  However, Meteor 1.4.2.7 is no longer the default recommended release so this should be less common now.  Still not out of the question though, but hopefully the 1.4.3.x series will be the easier upgrade path (and the 0.6.3 `ecmascript` is definitely the preferred implementation and I believe critical for Meteor 1.5 dynamic imports).

On that note, you're saying you're having problems with Meteor 1.4.3.1 and `ecmascript@0.6.3` so let's try to address that.  Most importantly, can you provide the reproduction you've suggested above?

**Comment by zvictor** [comment]
Thank you for spending your time explaining the process in details. Amazing community support! 🍻

---

I think I have a diagnose for it now, and a reproducible.

Basically the new version drops the support to [transform-export-extensions](https://babeljs.io/docs/plugins/transform-export-extensions/), which is part of Babel's Stage 1 preset.

Check it out at [zvictor/meteor-bug-8432](https://github.com/zvictor/meteor-bug-8432). There are 2 commits: in the first it is working (Meteor 1.4.2.7 with ecmascript@=0.6.1), in the last it is broken (Meteor 1.4.3.1).

**Comment by flippyhead** [comment]
FWIW I just upgraded a simple application to Meteor 1.4.3.1, it pinned ecmascript@0.6.3 and I have the same problem.

**Comment by MiMaMuh** [comment]
@flippyhead @zvictor same here, also using babel-plugin-transform-export-extensions with meteor 1.4.3.1 and ecmascript@0.6.3

**Comment by zvictor** [comment]
@abernix would you consider removing the `can't-reproduce` label? :)

**Comment by benjamn** [comment]
Although the general problem of running Babel plugins in the correct order is tricky and worth investigating further, in this case I honestly think we should just include [`transform-export-extensions`](https://www.npmjs.com/package/babel-plugin-transform-export-extensions) in [`babel-preset-meteor`](https://www.npmjs.com/package/babel-preset-meteor) by default. Those extensions are [still at stage 1](https://github.com/tc39/proposals#active-proposals), but they are not controversial as far as I can tell.

**Comment by benjamn** [comment]
This should be fixed if you run `meteor update --release 1.4.3.3-beta.3` in your application directory. Now that Reify supports the `export`-`from` extensions, you don't need the `transform-export-extensions` plugin anymore, though it shouldn't do any harm. Feel free to reopen this issue if it's not resolved for you.
